ZOOKEEPER-3156: Add in option to canonicalize host name

revans2 opened the pull request:
This is the master and 3.5 version of #648.  It should apply cleanly to both lines, but if you want a separate pull request for each I am happy to do it.

eolivelli commented on src/java/main/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -1102,7 +1103,32 @@ private void startConnect(InetSocketAddress addr) throws IOException {
+
+            boolean canonicalize = true;
+            try {
+                canonicalize = Boolean.parseBoolean(System.getProperty(ZKClientConfig.ZK_SASL_CLIENT_CANONICALIZE_HOSTNAME,
```
Isn't it clientConfig instead of System ?

revans2 replied:
Great point, missed that on the port.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2300/

eolivelli commented:
Looks great now

Like for branch-3.4 I think we'd better to have some test

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2301/

revans2 commented:
Added in the test.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2303/

lvfangmin commented on src/java/main/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -793,7 +794,87 @@ public RWServerFoundException(String msg) {
+        try {
+            canonicalize = Boolean.parseBoolean(clientConfig.getProperty(ZKClientConfig.ZK_SASL_CLIENT_CANONICALIZE_HOSTNAME,
+                ZKClientConfig.ZK_SASL_CLIENT_CANONICALIZE_HOSTNAME_DEFAULT));
+        } catch (IllegalArgumentException ea) {
```
Should we print the log when IllegalArgumentException throws?

revans2 replied:
The IllegalArgumentException is coming from the `Boolean.parseBoolean`.  If it cannot understand the value as either true or false it will throw the exception and we end up with it being true, the default.  If you want a log I am happy to add it, because "no" will be interpreted as true, which might be confusing.

anmolnar replied:
I don't see any of these methods (parseBoolean() and getProperty()) throw IllegalArgumentException, so you don't need to catch here.

In your original PR, you used System.getProperty() which throws it:
https://github.com/apache/zookeeper/pull/648/files#diff-ae38c175a23a47be52ae31ed1f1518a3R781

But this is master and we use ZKConfig here.

anmolnar replied:
`Boolean.parseBoolean()` returns true if the string is "true" (ignore case) and false otherwise. No exception thrown.

lvfangmin commented on src/java/main/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -793,7 +794,87 @@ public RWServerFoundException(String msg) {
     }
-    
+
+    static class MockableInetSocketAddress {
```
It's adding too much complexity to the code in order to test it, are we able to create the stub to extend the InetSocketAddress in the test case itself?

anmolnar replied:
I agree. Unfortunately methods of `InetSocketAddress` and `InetAddress` are final and cannot be mocked. There's a Mockito extension which could help, but it's quite recent and an opt-in feature:
https://github.com/mockito/mockito/wiki/What's-new-in-Mockito-2#mock-the-unmockable-opt-in-mocking-of-final-classesmethods

Implementing wrapper classes is good approach, but I wouldn't put them into `ClientCnxn.java`. I suggest creating a new file with wrapper classes (nit: renaming them to \*Wrapper instead of Mockable\* would be better) and also put `getServerPrincipal()` into this file as well.

anmolnar commented on src/java/test/org/apache/zookeeper/ClientCanonicalizeTest.java:
```diff
@@ -0,0 +1,59 @@
+import org.junit.Assert;
+import org.junit.Test;
+
+import static org.mockito.Mockito.*;
```
Please avoid using asterisk import.

anmolnar commented on src/java/test/org/apache/zookeeper/ClientCanonicalizeTest.java:
```diff
@@ -0,0 +1,59 @@
+
+        ZKClientConfig conf = new ZKClientConfig();
+        String principal = ClientCnxn.getServerPrincipal(addr, conf);
+        Assert.assertEquals("zookeeper/zk1.apache.org", principal);
```
Please add some comment to the assert to describe the scenario when it fails.

anmolnar commented on src/java/test/org/apache/zookeeper/ClientCanonicalizeTest.java:
```diff
@@ -0,0 +1,59 @@
+        ZKClientConfig conf = new ZKClientConfig();
+        conf.setProperty(ZKClientConfig.ZK_SASL_CLIENT_CANONICALIZE_HOSTNAME, "false");
+        String principal = ClientCnxn.getServerPrincipal(addr, conf);
+        Assert.assertEquals("zookeeper/zookeeper.apache.org", principal);
```
Same here.

anmolnar commented on src/java/main/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -793,7 +794,87 @@ public RWServerFoundException(String msg) {
+
+            String canonicalHostName = ia.getCanonicalHostName();
+            //avoid using literal IP address when security check fails
+            if (!canonicalHostName.equals(ia.getHostAddress())) {
```
Would you please add test to cover this code path: canonicalization is ON, but getCanonicalHostName() returns IP address.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2321/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2325/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2324/

revans2 commented:
Ping @eolivelli @lvfangmin @anmolnar 

I think I have addressed all of the concerns on both branches.  The test failure here looks unrelated to the changes that I made.

revans2 commented:
Thanks for all of the reviews I just rebased and squashed commits.

anmolnar commented:
@lvfangmin are u happy with the change?

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2348/

revans2 commented:
I just rebased to deal with the directories being moved around.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2371/

anmolnar commented:
@revans2 We might need another committer to take a look.
Also you probably need to trigger another build to get a green one.

eolivelli commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2459/

anmolnar commented:
@revans2 Please rebase.

revans2 commented:
@anmolnar sorry it took me so long.  I was out on vacation.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2549/

anmolnar commented:
Merged to 3.5 and master branches. Thanks @revans2 !
